File: apps/fluster/lib/features/command_palette/data/models/command_palette_category.dart
import 'package:async_redux/async_redux.dart';
import 'package:fluster/features/command_palette/data/models/command_palette_entry.dart';
import 'package:fluster/features/command_palette/state/actions/append_command_palette_category.dart';
import 'package:fluster/features/command_palette/state/command_palette_state.dart';
import 'package:flutter/material.dart';

abstract class CommandPaletteCategory extends CommandPaletteEntry {
  final CommandPaletteView layout;
  const CommandPaletteCategory({
    required super.label,
    required super.category,
    required super.desc,
    this.layout = CommandPaletteView.list,
  });

  Future<void> onEnter() async {}
  Future<List<CommandPaletteEntry>> getItemsOnEnter();
  Future<List<CommandPaletteEntry>> getItemsOnQueryChange(String query);

  IconData getIcon();

  @override
  void callAction(context) {
    context.dispatch(AppendCommandPaletteCategoryAction(this));
  }
}

File: apps/fluster/src/features/command_palette/data/models/command_palette_category.ts
import { Location } from "react-router";
import { CommandPaletteAnyEntry } from "./command_palette_any_entry";
import { ReactNode } from "react";
import { CommandPaletteState } from "#/command_palette/state/command_palette_provider";

export abstract class CommandPaletteCategory extends CommandPaletteAnyEntry {
  constructor(label: string, id: string) {
    super(label, id);
  }
  abstract filterByLocation(location: Location): boolean;
  abstract getItems(location: Location): Promise<CommandPaletteAnyEntry[]>;
  abstract bottomBar(state: CommandPaletteState): ReactNode;
}
